File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/MaterialTools.cs
using System;
using System.Collections.Generic;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Utility class for Material tool operations
    /// </summary>
    public static class MaterialToolUtils
    {
        /// <summary>
        /// Get the default lit shader based on the current render pipeline
        /// </summary>
        public static string GetDefaultShaderName()
        {
            // Check for URP
            if (UnityEngine.Rendering.GraphicsSettings.currentRenderPipeline != null)
            {
                string pipelineName = UnityEngine.Rendering.GraphicsSettings.currentRenderPipeline.GetType().Name;

                if (pipelineName.Contains("Universal") || pipelineName.Contains("URP"))
                {
                    return "Universal Render Pipeline/Lit";
                }
                else if (pipelineName.Contains("HD") || pipelineName.Contains("HDRP"))
                {
                    return "HDRP/Lit";
                }
                else
                {
                    McpLogger.LogWarning("Unknown render pipeline, defaulting to Standard shader");
                }
            }

            // Default to Standard (Built-in Render Pipeline)
            return "Standard";
        }

        /// <summary>
        /// Find a shader by name, searching common Unity shader paths
        /// </summary>
        public static Shader FindShader(string shaderName)
        {
            // Try direct lookup first
            Shader shader = Shader.Find(shaderName);
            if (shader != null)
            {
                return shader;
            }

            // Common shader path prefixes to try
            string[] prefixes = new string[]
            {
                "",
                "Standard",
                "Universal Render Pipeline/",
                "URP/",
                "HDRP/",
                "Hidden/",
                "Legacy Shaders/",
                "Mobile/",
                "Particles/",
                "Skybox/",
                "Sprites/",
                "UI/",
                "Unlit/"
            };

            foreach (string prefix in prefixes)
            {
                shader = Shader.Find(prefix + shaderName);
                if (shader != null)
                {
                    return shader;
                }
            }

            return null;
        }

        /// <summary>
        /// Load a material from an asset path
        /// </summary>
        public static Material LoadMaterial(string materialPath)
        {
            if (string.IsNullOrEmpty(materialPath))
            {
                return null;
            }

            // Ensure path starts with Assets/
            if (!materialPath.StartsWith("Assets/"))
            {
                materialPath = "Assets/" + materialPath;
            }

            // Ensure .mat extension
            if (!materialPath.EndsWith(".mat"))
            {
                materialPath += ".mat";
            }

            return AssetDatabase.LoadAssetAtPath<Material>(materialPath);
        }

        /// <summary>
        /// Convert a JToken to a shader property value
        /// </summary>
        public static object ConvertPropertyValue(JToken token, ShaderUtil.ShaderPropertyType propertyType)
        {
            if (token == null)
            {
                return null;
            }

            switch (propertyType)
            {
                case ShaderUtil.ShaderPropertyType.Color:
                    if (token.Type == JTokenType.Object)
                    {
                        JObject color = (JObject)token;
                        return new Color(
                            color["r"]?.ToObject<float>() ?? 0f,
                            color["g"]?.ToObject<float>() ?? 0f,
                            color["b"]?.ToObject<float>() ?? 0f,
                            color["a"]?.ToObject<float>() ?? 1f
                        );
                    }
                    break;

                case ShaderUtil.ShaderPropertyType.Vector:
                    if (token.Type == JTokenType.Object)
                    {
                        JObject vec = (JObject)token;
                        return new Vector4(
                            vec["x"]?.ToObject<float>() ?? 0f,
                            vec["y"]?.ToObject<float>() ?? 0f,
                            vec["z"]?.ToObject<float>() ?? 0f,
                            vec["w"]?.ToObject<float>() ?? 0f
                        );
                    }
                    break;

                case ShaderUtil.ShaderPropertyType.Float:
                case ShaderUtil.ShaderPropertyType.Range:
                    return token.ToObject<float>();

                case ShaderUtil.ShaderPropertyType.TexEnv:
                    // Texture path
                    string texPath = token.ToObject<string>();
                    if (!string.IsNullOrEmpty(texPath))
                    {
                        if (!texPath.StartsWith("Assets/"))
                        {
                            texPath = "Assets/" + texPath;
                        }
                        return AssetDatabase.LoadAssetAtPath<Texture>(texPath);
                    }
                    break;

                case ShaderUtil.ShaderPropertyType.Int:
                    return token.ToObject<int>();
            }

            return null;
        }

        /// <summary>
        /// Find a GameObject by instance ID or path
        /// </summary>
        public static GameObject FindGameObject(int? instanceId, string objectPath)
        {
            GameObject gameObject = null;

            if (instanceId.HasValue)
            {
                gameObject = EditorUtility.InstanceIDToObject(instanceId.Value) as GameObject;
            }
            else if (!string.IsNullOrEmpty(objectPath))
            {
                gameObject = GameObject.Find(objectPath);

                if (gameObject == null)
                {
                    // Try to find using hierarchy path
                    gameObject = FindGameObjectByPath(objectPath);
                }
            }

            return gameObject;
        }

        /// <summary>
        /// Find a GameObject by its hierarchy path
        /// </summary>
        private static GameObject FindGameObjectByPath(string path)
        {
            string[] pathParts = path.Split('/');
            GameObject[] rootGameObjects = UnityEngine.SceneManagement.SceneManager.GetActiveScene().GetRootGameObjects();

            if (pathParts.Length == 0)
            {
                return null;
            }

            foreach (GameObject rootObj in rootGameObjects)
            {
                if (rootObj.name == pathParts[0])
                {
                    GameObject current = rootObj;

                    for (int i = 1; i < pathParts.Length; i++)
                    {
                        Transform child = current.transform.Find(pathParts[i]);
                        if (child == null)
                        {
                            return null;
                        }
                        current = child.gameObject;
                    }

                    return current;
                }
            }

            return null;
        }
    }

    /// <summary>
    /// Tool for creating new materials
    /// </summary>
    public class CreateMaterialTool : McpToolBase
    {
        public CreateMaterialTool()
        {
            Name = "create_material";
            Description = "Creates a new material with the specified shader and saves it to the project";
        }

        public override JObject Execute(JObject parameters)
        {
            // Extract parameters
            string name = parameters["name"]?.ToObject<string>();
            string shaderName = parameters["shader"]?.ToObject<string>();
            string savePath = parameters["savePath"]?.ToObject<string>();
            JObject properties = parameters["properties"] as JObject;
            JObject colorParam = parameters["color"] as JObject;

            // Validate required parameters
            if (string.IsNullOrEmpty(name))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'name' not provided",
                    "validation_error"
                );
            }

            if (string.IsNullOrEmpty(savePath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'savePath' not provided",
                    "validation_error"
                );
            }

            // Default shader based on current render pipeline
            if (string.IsNullOrEmpty(shaderName))
            {
                shaderName = MaterialToolUtils.GetDefaultShaderName();
            }

            // Find the shader
            Shader shader = MaterialToolUtils.FindShader(shaderName);
            if (shader == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Shader '{shaderName}' not found in Unity",
                    "not_found_error"
                );
            }

            // Ensure save path has proper format
            if (!savePath.StartsWith("Assets/"))
            {
                savePath = "Assets/" + savePath;
            }
            if (!savePath.EndsWith(".mat"))
            {
                savePath += ".mat";
            }

            // Ensure directory exists
            string directory = System.IO.Path.GetDirectoryName(savePath);
            if (!System.IO.Directory.Exists(directory))
            {
                System.IO.Directory.CreateDirectory(directory);
            }

            // Create the material
            Material material = new Material(shader);
            material.name = name;

            // Apply color if provided (auto-detect correct property name)
            if (colorParam != null)
            {
                Color color = new Color(
                    colorParam["r"]?.ToObject<float>() ?? 1f,
                    colorParam["g"]?.ToObject<float>() ?? 1f,
                    colorParam["b"]?.ToObject<float>() ?? 1f,
                    colorParam["a"]?.ToObject<float>() ?? 1f
                );
                ApplyBaseColor(material, color);
            }

            // Apply initial properties if provided
            if (properties != null && properties.Count > 0)
            {
                ApplyMaterialProperties(material, properties);
            }

            // Save the material as an asset
            AssetDatabase.CreateAsset(material, savePath);
            AssetDatabase.SaveAssets();
            AssetDatabase.Refresh();

            McpLogger.LogInfo($"[MCP Unity] Created material '{name}' with shader '{shaderName}' at '{savePath}'");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully created material '{name}' with shader '{shaderName}'",
                ["materialPath"] = savePath,
                ["materialName"] = name,
                ["shaderName"] = shader.name
            };
        }

        private void ApplyMaterialProperties(Material material, JObject properties)
        {
            Shader shader = material.shader;
            int propertyCount = ShaderUtil.GetPropertyCount(shader);

            foreach (var prop in properties.Properties())
            {
                string propName = prop.Name;
                JToken propValue = prop.Value;

                // Find the property in the shader
                for (int i = 0; i < propertyCount; i++)
                {
                    string shaderPropName = ShaderUtil.GetPropertyName(shader, i);
                    if (shaderPropName == propName)
                    {
                        ShaderUtil.ShaderPropertyType propType = ShaderUtil.GetPropertyType(shader, i);
                        object value = MaterialToolUtils.ConvertPropertyValue(propValue, propType);

                        if (value != null)
                        {
                            SetMaterialProperty(material, propName, propType, value);
                        }
                        break;
                    }
                }
            }
        }

        private void SetMaterialProperty(Material material, string propName, ShaderUtil.ShaderPropertyType propType, object value)
        {
            switch (propType)
            {
                case ShaderUtil.ShaderPropertyType.Color:
                    material.SetColor(propName, (Color)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Vector:
                    material.SetVector(propName, (Vector4)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Float:
                case ShaderUtil.ShaderPropertyType.Range:
                    material.SetFloat(propName, (float)value);
                    break;
                case ShaderUtil.ShaderPropertyType.TexEnv:
                    material.SetTexture(propName, (Texture)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Int:
                    material.SetInt(propName, (int)value);
                    break;
            }
        }

        /// <summary>
        /// Apply base color to material, auto-detecting the correct property name
        /// </summary>
        private void ApplyBaseColor(Material material, Color color)
        {
            // Common color property names in order of preference
            string[] colorPropertyNames = new string[]
            {
                "_BaseColor",    // URP Lit, HDRP Lit
                "_Color",        // Standard, Legacy shaders
                "_TintColor",    // Particle shaders
                "_MainColor"     // Some custom shaders
            };

            foreach (string propName in colorPropertyNames)
            {
                if (material.HasProperty(propName))
                {
                    material.SetColor(propName, color);
                    return;
                }
            }

            // Fallback: try to find any color property
            Shader shader = material.shader;
            int propertyCount = ShaderUtil.GetPropertyCount(shader);
            for (int i = 0; i < propertyCount; i++)
            {
                if (ShaderUtil.GetPropertyType(shader, i) == ShaderUtil.ShaderPropertyType.Color)
                {
                    string propName = ShaderUtil.GetPropertyName(shader, i);
                    material.SetColor(propName, color);
                    return;
                }
            }
        }
    }

    /// <summary>
    /// Tool for assigning materials to GameObjects
    /// </summary>
    public class AssignMaterialTool : McpToolBase
    {
        public AssignMaterialTool()
        {
            Name = "assign_material";
            Description = "Assigns a material to a GameObject's Renderer component";
        }

        public override JObject Execute(JObject parameters)
        {
            // Extract parameters
            int? instanceId = parameters["instanceId"]?.ToObject<int?>();
            string objectPath = parameters["objectPath"]?.ToObject<string>();
            string materialPath = parameters["materialPath"]?.ToObject<string>();
            int slot = parameters["slot"]?.ToObject<int?>() ?? 0;

            // Validate parameters
            if (!instanceId.HasValue && string.IsNullOrEmpty(objectPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Either 'instanceId' or 'objectPath' must be provided",
                    "validation_error"
                );
            }

            if (string.IsNullOrEmpty(materialPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'materialPath' not provided",
                    "validation_error"
                );
            }

            // Find the GameObject
            GameObject gameObject = MaterialToolUtils.FindGameObject(instanceId, objectPath);
            if (gameObject == null)
            {
                string identifier = instanceId.HasValue ? $"ID {instanceId.Value}" : $"path '{objectPath}'";
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"GameObject with {identifier} not found",
                    "not_found_error"
                );
            }

            // Get the Renderer component
            Renderer renderer = gameObject.GetComponent<Renderer>();
            if (renderer == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"GameObject '{gameObject.name}' does not have a Renderer component",
                    "component_error"
                );
            }

            // Load the material
            Material material = MaterialToolUtils.LoadMaterial(materialPath);
            if (material == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Material at path '{materialPath}' not found",
                    "not_found_error"
                );
            }

            // Validate slot index
            Material[] materials = renderer.sharedMaterials;
            if (slot < 0 || slot >= materials.Length)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Material slot {slot} is out of range. GameObject has {materials.Length} material slot(s) (0-{materials.Length - 1})",
                    "validation_error"
                );
            }

            // Record for undo
            Undo.RecordObject(renderer, $"Assign Material to {gameObject.name}");

            // Assign the material
            materials[slot] = material;
            renderer.sharedMaterials = materials;

            // Mark as dirty
            EditorUtility.SetDirty(renderer);
            if (PrefabUtility.IsPartOfAnyPrefab(gameObject))
            {
                PrefabUtility.RecordPrefabInstancePropertyModifications(renderer);
            }

            McpLogger.LogInfo($"[MCP Unity] Assigned material '{material.name}' to '{gameObject.name}' at slot {slot}");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully assigned material '{material.name}' to '{gameObject.name}' at slot {slot}",
                ["gameObjectName"] = gameObject.name,
                ["materialName"] = material.name,
                ["slot"] = slot
            };
        }
    }

    /// <summary>
    /// Tool for modifying material properties
    /// </summary>
    public class ModifyMaterialTool : McpToolBase
    {
        public ModifyMaterialTool()
        {
            Name = "modify_material";
            Description = "Modifies properties of an existing material";
        }

        public override JObject Execute(JObject parameters)
        {
            // Extract parameters
            string materialPath = parameters["materialPath"]?.ToObject<string>();
            JObject properties = parameters["properties"] as JObject;

            // Validate parameters
            if (string.IsNullOrEmpty(materialPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'materialPath' not provided",
                    "validation_error"
                );
            }

            if (properties == null || properties.Count == 0)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'properties' not provided or empty",
                    "validation_error"
                );
            }

            // Load the material
            Material material = MaterialToolUtils.LoadMaterial(materialPath);
            if (material == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Material at path '{materialPath}' not found",
                    "not_found_error"
                );
            }

            // Record for undo
            Undo.RecordObject(material, $"Modify Material {material.name}");

            // Apply properties
            Shader shader = material.shader;
            int propertyCount = ShaderUtil.GetPropertyCount(shader);
            List<string> modifiedProperties = new List<string>();
            List<string> unknownProperties = new List<string>();

            foreach (var prop in properties.Properties())
            {
                string propName = prop.Name;
                JToken propValue = prop.Value;
                bool found = false;

                // Find the property in the shader
                for (int i = 0; i < propertyCount; i++)
                {
                    string shaderPropName = ShaderUtil.GetPropertyName(shader, i);
                    if (shaderPropName == propName)
                    {
                        found = true;
                        ShaderUtil.ShaderPropertyType propType = ShaderUtil.GetPropertyType(shader, i);
                        object value = MaterialToolUtils.ConvertPropertyValue(propValue, propType);

                        if (value != null)
                        {
                            SetMaterialProperty(material, propName, propType, value);
                            modifiedProperties.Add(propName);
                        }
                        break;
                    }
                }

                if (!found)
                {
                    unknownProperties.Add(propName);
                }
            }

            // Mark as dirty and save
            EditorUtility.SetDirty(material);
            AssetDatabase.SaveAssets();
            AssetDatabase.Refresh();

            McpLogger.LogInfo($"[MCP Unity] Modified material '{material.name}': {string.Join(", ", modifiedProperties)}");

            JObject result = new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully modified material '{material.name}'",
                ["materialName"] = material.name,
                ["modifiedProperties"] = new JArray(modifiedProperties)
            };

            if (unknownProperties.Count > 0)
            {
                result["unknownProperties"] = new JArray(unknownProperties);
                result["message"] = $"Modified material '{material.name}'. Some properties were not found: {string.Join(", ", unknownProperties)}";
            }

            return result;
        }

        private void SetMaterialProperty(Material material, string propName, ShaderUtil.ShaderPropertyType propType, object value)
        {
            switch (propType)
            {
                case ShaderUtil.ShaderPropertyType.Color:
                    material.SetColor(propName, (Color)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Vector:
                    material.SetVector(propName, (Vector4)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Float:
                case ShaderUtil.ShaderPropertyType.Range:
                    material.SetFloat(propName, (float)value);
                    break;
                case ShaderUtil.ShaderPropertyType.TexEnv:
                    material.SetTexture(propName, (Texture)value);
                    break;
                case ShaderUtil.ShaderPropertyType.Int:
                    material.SetInt(propName, (int)value);
                    break;
            }
        }
    }

    /// <summary>
    /// Tool for getting material information
    /// </summary>
    public class GetMaterialInfoTool : McpToolBase
    {
        public GetMaterialInfoTool()
        {
            Name = "get_material_info";
            Description = "Gets detailed information about a material including its shader and all properties";
        }

        public override JObject Execute(JObject parameters)
        {
            // Extract parameters
            string materialPath = parameters["materialPath"]?.ToObject<string>();

            // Validate parameters
            if (string.IsNullOrEmpty(materialPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'materialPath' not provided",
                    "validation_error"
                );
            }

            // Load the material
            Material material = MaterialToolUtils.LoadMaterial(materialPath);
            if (material == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Material at path '{materialPath}' not found",
                    "not_found_error"
                );
            }

            // Get shader info
            Shader shader = material.shader;
            int propertyCount = ShaderUtil.GetPropertyCount(shader);

            // Build properties array
            JArray propertiesArray = new JArray();
            for (int i = 0; i < propertyCount; i++)
            {
                string propName = ShaderUtil.GetPropertyName(shader, i);
                string propDescription = ShaderUtil.GetPropertyDescription(shader, i);
                ShaderUtil.ShaderPropertyType propType = ShaderUtil.GetPropertyType(shader, i);

                JObject propInfo = new JObject
                {
                    ["name"] = propName,
                    ["description"] = propDescription,
                    ["type"] = propType.ToString()
                };

                // Get current value
                propInfo["value"] = GetPropertyValue(material, propName, propType);

                // Add range info if applicable
                if (propType == ShaderUtil.ShaderPropertyType.Range)
                {
                    propInfo["rangeMin"] = ShaderUtil.GetRangeLimits(shader, i, 1);
                    propInfo["rangeMax"] = ShaderUtil.GetRangeLimits(shader, i, 2);
                }

                propertiesArray.Add(propInfo);
            }

            // Build render queue info
            string renderQueueName = "Custom";
            int renderQueue = material.renderQueue;
            if (renderQueue <= 2000) renderQueueName = "Background";
            else if (renderQueue <= 2450) renderQueueName = "Geometry";
            else if (renderQueue <= 2500) renderQueueName = "AlphaTest";
            else if (renderQueue <= 3000) renderQueueName = "Transparent";
            else renderQueueName = "Overlay";

            McpLogger.LogInfo($"[MCP Unity] Retrieved info for material '{material.name}'");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Material info for '{material.name}'",
                ["materialName"] = material.name,
                ["materialPath"] = materialPath,
                ["shaderName"] = shader.name,
                ["renderQueue"] = renderQueue,
                ["renderQueueCategory"] = renderQueueName,
                ["enableInstancing"] = material.enableInstancing,
                ["doubleSidedGI"] = material.doubleSidedGI,
                ["passCount"] = material.passCount,
                ["properties"] = propertiesArray
            };
        }

        private JToken GetPropertyValue(Material material, string propName, ShaderUtil.ShaderPropertyType propType)
        {
            switch (propType)
            {
                case ShaderUtil.ShaderPropertyType.Color:
                    Color color = material.GetColor(propName);
                    return new JObject
                    {
                        ["r"] = color.r,
                        ["g"] = color.g,
                        ["b"] = color.b,
                        ["a"] = color.a
                    };

                case ShaderUtil.ShaderPropertyType.Vector:
                    Vector4 vec = material.GetVector(propName);
                    return new JObject
                    {
                        ["x"] = vec.x,
                        ["y"] = vec.y,
                        ["z"] = vec.z,
                        ["w"] = vec.w
                    };

                case ShaderUtil.ShaderPropertyType.Float:
                case ShaderUtil.ShaderPropertyType.Range:
                    return material.GetFloat(propName);

                case ShaderUtil.ShaderPropertyType.TexEnv:
                    Texture tex = material.GetTexture(propName);
                    if (tex != null)
                    {
                        return AssetDatabase.GetAssetPath(tex);
                    }
                    return null;

                case ShaderUtil.ShaderPropertyType.Int:
                    return material.GetInt(propName);

                default:
                    return null;
            }
        }
    }
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/materialTools.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Color schema for material properties
const colorSchema = z.object({
  r: z.number().min(0).max(1).describe('Red component (0-1)'),
  g: z.number().min(0).max(1).describe('Green component (0-1)'),
  b: z.number().min(0).max(1).describe('Blue component (0-1)'),
  a: z.number().min(0).max(1).optional().default(1).describe('Alpha component (0-1)')
});

// Vector4 schema for material properties
const vector4Schema = z.object({
  x: z.number().describe('X component'),
  y: z.number().describe('Y component'),
  z: z.number().describe('Z component'),
  w: z.number().optional().default(0).describe('W component')
});

// ============================================================================
// CREATE MATERIAL TOOL
// ============================================================================

const createMaterialToolName = 'create_material';
const createMaterialToolDescription = 'Creates a new material with the specified shader and saves it to the project. Use the "color" parameter for easy color setting.';
const createMaterialParamsSchema = z.object({
  name: z.string().describe('The name of the material'),
  shader: z.string().optional().describe('The shader name. Auto-detects render pipeline if not specified (URP: "Universal Render Pipeline/Lit", Built-in: "Standard")'),
  savePath: z.string().describe('The asset path to save the material (e.g., "Assets/Materials/MyMaterial.mat")'),
  color: colorSchema.optional().describe('The base color of the material. Auto-detects correct property name (_BaseColor for URP, _Color for Standard)'),
  properties: z.record(z.any()).optional().describe('Optional initial property values as key-value pairs (advanced usage)')
});

/**
 * Registers the Create Material tool with the MCP server
 */
export function registerCreateMaterialTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${createMaterialToolName}`);

  server.tool(
    createMaterialToolName,
    createMaterialToolDescription,
    createMaterialParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${createMaterialToolName}`, params);
        const result = await createMaterialHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${createMaterialToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${createMaterialToolName}`, error);
        throw error;
      }
    }
  );
}

async function createMaterialHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.name) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'name' must be provided"
    );
  }

  if (!params.savePath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'savePath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: createMaterialToolName,
    params: {
      name: params.name,
      shader: params.shader, // Let Unity auto-detect if not specified
      savePath: params.savePath,
      color: params.color,   // Auto-maps to correct shader property (_BaseColor, _Color, etc.)
      properties: params.properties
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to create material'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully created material '${params.name}'`
    }]
  };
}

// ============================================================================
// ASSIGN MATERIAL TOOL
// ============================================================================

const assignMaterialToolName = 'assign_material';
const assignMaterialToolDescription = 'Assigns a material to a GameObject\'s Renderer component at a specific material slot';
const assignMaterialParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
  materialPath: z.string().describe('The asset path to the material (e.g., "Assets/Materials/MyMaterial.mat")'),
  slot: z.number().int().min(0).optional().default(0).describe('The material slot index (default: 0)')
});

/**
 * Registers the Assign Material tool with the MCP server
 */
export function registerAssignMaterialTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${assignMaterialToolName}`);

  server.tool(
    assignMaterialToolName,
    assignMaterialToolDescription,
    assignMaterialParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${assignMaterialToolName}`, params);
        const result = await assignMaterialHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${assignMaterialToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${assignMaterialToolName}`, error);
        throw error;
      }
    }
  );
}

async function assignMaterialHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if ((params.instanceId === undefined || params.instanceId === null) &&
      (!params.objectPath || params.objectPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'instanceId' or 'objectPath' must be provided"
    );
  }

  if (!params.materialPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'materialPath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: assignMaterialToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      materialPath: params.materialPath,
      slot: params.slot ?? 0
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to assign material'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully assigned material`
    }]
  };
}

// ============================================================================
// MODIFY MATERIAL TOOL
// ============================================================================

const modifyMaterialToolName = 'modify_material';
const modifyMaterialToolDescription = 'Modifies properties of an existing material. Supports colors (e.g., _Color), floats (e.g., _Metallic), and textures (e.g., _MainTex path)';
const modifyMaterialParamsSchema = z.object({
  materialPath: z.string().describe('The asset path to the material (e.g., "Assets/Materials/MyMaterial.mat")'),
  properties: z.record(z.any()).describe('Property name to value mapping. Colors: {r,g,b,a}, Vectors: {x,y,z,w}, Floats: number, Textures: asset path string')
});

/**
 * Registers the Modify Material tool with the MCP server
 */
export function registerModifyMaterialTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${modifyMaterialToolName}`);

  server.tool(
    modifyMaterialToolName,
    modifyMaterialToolDescription,
    modifyMaterialParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${modifyMaterialToolName}`, params);
        const result = await modifyMaterialHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${modifyMaterialToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${modifyMaterialToolName}`, error);
        throw error;
      }
    }
  );
}

async function modifyMaterialHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.materialPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'materialPath' must be provided"
    );
  }

  if (!params.properties || Object.keys(params.properties).length === 0) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'properties' must be provided and contain at least one property"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: modifyMaterialToolName,
    params: {
      materialPath: params.materialPath,
      properties: params.properties
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to modify material'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully modified material`
    }]
  };
}

// ============================================================================
// GET MATERIAL INFO TOOL
// ============================================================================

const getMaterialInfoToolName = 'get_material_info';
const getMaterialInfoToolDescription = 'Gets detailed information about a material including its shader and all properties with current values';
const getMaterialInfoParamsSchema = z.object({
  materialPath: z.string().describe('The asset path to the material (e.g., "Assets/Materials/MyMaterial.mat")')
});

/**
 * Registers the Get Material Info tool with the MCP server
 */
export function registerGetMaterialInfoTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${getMaterialInfoToolName}`);

  server.tool(
    getMaterialInfoToolName,
    getMaterialInfoToolDescription,
    getMaterialInfoParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${getMaterialInfoToolName}`, params);
        const result = await getMaterialInfoHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${getMaterialInfoToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${getMaterialInfoToolName}`, error);
        throw error;
      }
    }
  );
}

async function getMaterialInfoHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.materialPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'materialPath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: getMaterialInfoToolName,
    params: {
      materialPath: params.materialPath
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get material info'
    );
  }

  // Format the response with material details
  let text = `Material: ${response.materialName}\n`;
  text += `Shader: ${response.shaderName}\n`;
  text += `Render Queue: ${response.renderQueue} (${response.renderQueueCategory})\n`;
  text += `Instancing: ${response.enableInstancing}\n`;
  text += `Pass Count: ${response.passCount}\n\n`;
  text += `Properties:\n`;

  if (response.properties && Array.isArray(response.properties)) {
    for (const prop of response.properties) {
      let valueStr = '';
      if (prop.value === null || prop.value === undefined) {
        valueStr = 'null';
      } else if (typeof prop.value === 'object') {
        valueStr = JSON.stringify(prop.value);
      } else {
        valueStr = String(prop.value);
      }

      text += `  ${prop.name} (${prop.type}): ${valueStr}`;
      if (prop.description) {
        text += ` - ${prop.description}`;
      }
      text += '\n';
    }
  }

  return {
    content: [{
      type: 'text',
      text: text
    }],
    data: {
      materialName: response.materialName,
      materialPath: response.materialPath,
      shaderName: response.shaderName,
      renderQueue: response.renderQueue,
      renderQueueCategory: response.renderQueueCategory,
      enableInstancing: response.enableInstancing,
      doubleSidedGI: response.doubleSidedGI,
      passCount: response.passCount,
      properties: response.properties
    }
  };
}
